src/main/scala/no/ndla/frontpageapi/service/WriteService.scala

```diff
@@ -19,26 +19,51 @@ trait WriteService {
 
   class WriteService {
 
-    def newSubjectPage(subject: api.NewOrUpdateSubjectFrontPageData): Try[api.SubjectPageData] = {
+    def newSubjectPage(subject: api.NewSubjectFrontPageData): Try[api.SubjectPageData] = {
       for {
         convertedSubject <- ConverterService.toDomainSubjectPage(subject)
         subjectPage <- subjectPageRepository.newSubjectPage(convertedSubject, subject.externalId)
-      } yield ConverterService.toApiSubjectPage(subjectPage, "nb")
+        converted <- ConverterService.toApiSubjectPage(subjectPage, "nb")
+      } yield converted
     }
 
-    def updateSubjectPage(id: Long, subject: api.NewOrUpdateSubjectFrontPageData): Try[api.SubjectPageData] = {
+    def updateSubjectPage(id: Long, subject: api.NewSubjectFrontPageData): Try[api.SubjectPageData] = {
       subjectPageRepository.exists(id) match {
         case Success(exists) if exists =>
           for {
             domainSubject <- ConverterService.toDomainSubjectPage(id, subject)
             subjectPage <- subjectPageRepository.updateSubjectPage(domainSubject)
-          } yield ConverterService.toApiSubjectPage(subjectPage, "nb")
+            converted <- ConverterService.toApiSubjectPage(subjectPage, "nb")
+          } yield converted
         case Success(_) =>
           Failure(NotFoundException(id))
         case Failure(ex) => Failure(ex)
       }
     }
 
+    def updateSubjectPage(id: Long, subject: api.UpdatedSubjectFrontPageData): Try[api.SubjectPageData] = {
+      subjectPageRepository.withId(id) match {
+        case Some(existingSubject) =>
+          for {
+            domainSubject <- ConverterService.toDomainSubjectPage(existingSubject, subject)
+            subjectPage <- subjectPageRepository.updateSubjectPage(domainSubject)
+            converted <- ConverterService.toApiSubjectPage(subjectPage, subject.about.get.head.language)
+          } yield converted
+        case None if subjectPageRepository.exists(id).getOrElse(false) =>
+          newFromUpdatedSubjectPage(subject) match {
+            case Failure(ex)             => Failure(ex)
+            case Success(newSubjectPage) => updateSubjectPage(id, newSubjectPage)
+          }
+        case None =>
+          Failure(NotFoundException(404))
+      }
+    }
+
+    private def newFromUpdatedSubjectPage(
+        updatedSubjectPage: api.UpdatedSubjectFrontPageData): Try[api.NewSubjectFrontPageData] = {
+      Try(updatedSubjectPage.asInstanceOf[api.NewSubjectFrontPageData])
```

Comment:
Det her så ut som en skummel cast. Må nok kalle konstruktøren i `api.NewSubjectFrontPagedata` manuelt med feltene fra `updatedSubjectPage` om vi trenger denne konverteringen :smile:

Reply:
Jaa skjønner. Men spørsmålet er som du sier om vi trenger den i det hele tatt da, for trenger den hvis en id kan eksistere i databasen uten at den har noe innhold, sant? For da må den heller lage en 'ny' side i stedet for å merge med en eksisterende?

Reply:
Japp riktig, den måten du gjorde det på nå likte jeg mye bedre :+1: :smile: